internal/agent/device/bootstrap.go

```diff
@@ -0,0 +1,295 @@
+
+	enrollmentCSR []byte
+	log           *logrus.Logger
+	// The log prefix used for testing
+	logPrefix string
+}
+
+func NewBootstrap(
+	deviceName string,
+	deviceWriter *fileio.Writer,
+	deviceReader *fileio.Reader,
+	enrollmentCSR []byte,
+	statusCollector status.Collector,
+	enrollmentClient *client.Enrollment,
+	managementEndpoint string,
+	enrollmentUIEndpoint string,
+	caFile string,
+	keyFile string,
+	managementGeneratedCertFile string,
+	backoff wait.Backoff,
+	log *logrus.Logger,
+	logPrefix string,
+) *Bootstrap {
+	return &Bootstrap{
+		deviceName:                  deviceName,
+		deviceWriter:                deviceWriter,
+		deviceReader:                deviceReader,
+		enrollmentCSR:               enrollmentCSR,
+		statusCollector:             statusCollector,
+		enrollmentClient:            enrollmentClient,
+		enrollmentUIEndpoint:        enrollmentUIEndpoint,
+		managementEndpoint:          managementEndpoint,
+		managementGeneratedCertFile: managementGeneratedCertFile,
+		caFile:                      caFile,
+		keyFile:                     keyFile,
+		log:                         log,
+		logPrefix:                   logPrefix,
+		backoff:                     backoff,
+	}
+}
+
+func (b *Bootstrap) Initialize(ctx context.Context) error {
+	b.log.Infof("%sbootstrapping device", b.logPrefix)
+	if err := b.ensureEnrollment(ctx); err != nil {
+		return err
+	}
+
+	if err := b.ensureRenderedSpec(ctx); err != nil {
+		return err
+	}
+
+	b.log.Infof("%sbootstrap complete", b.logPrefix)
+	return nil
+}
+
+func (b *Bootstrap) ensureRenderedSpec(ctx context.Context) error {
+	if err := b.deviceReader.CheckPathExists(b.managementGeneratedCertFile); err != nil {
+		return fmt.Errorf("generated cert: %q: %w", b.managementGeneratedCertFile, err)
+	}
+
+	// create the management client
+	managementHTTPClient, err := client.NewWithResponses(b.managementEndpoint, b.caFile, b.managementGeneratedCertFile, b.keyFile)
+	if err != nil {
+		return fmt.Errorf("create management client: %w", err)
+	}
+	managementClient := client.NewManagement(managementHTTPClient)
+
+	_, err = spec.EnsureDesiredRenderedSpec(ctx, b.log, b.logPrefix, b.deviceWriter, b.deviceReader, managementClient, b.deviceName, b.desiredRenderedFile, b.backoff)
+	if err != nil {
+		return fmt.Errorf("ensure desired rendered spec: %w", err)
+	}
+
+	_, err = spec.EnsureCurrentRenderedSpec(ctx, b.log, b.logPrefix, b.deviceWriter, b.deviceReader, b.deviceName, b.currentRenderedFile)
+	if err != nil {
+		return fmt.Errorf("ensure current rendered spec: %w", err)
+	}
+
+	return nil
+}
+
+func (b *Bootstrap) ensureEnrollment(ctx context.Context) error {
+	if !b.isBootstrapComplete() {
+		if err := b.writeEnrollmentBanner(); err != nil {
+			return err
+		}
+
+		if err := b.enrollmentRequest(ctx); err != nil {
+			return err
+		}
+
+		b.log.Infof("%swaiting for enrollment to be approved", b.logPrefix)
+		err := wait.ExponentialBackoff(b.backoff, func() (bool, error) {
```

Comment:
Ah, this is the reason the agent won't exit on shutdown:

https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#ExponentialBackoffWithContext

We need to use ExponentialBackoffWithContext here